Core state downloader will now retry indefinitely

RagnarW opened the pull request:
When the raft message handler need a new snapshot it will schedule
this on the downloader service. This will pause the application process
until download is successful.

RagnarW committed:
Core state downloader will now retry indefinitely

When the raft message handler need a new snapshot it will schedule
this on the downloader service. This will pause the application process
until download is successful.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/CoreServerModule.java:
```diff
@@ -177,10 +178,13 @@ private OnlineBackupKernelExtension pickBackupExtension( NeoStoreDataSource data
-                snapshotService, commandApplicationProcess, topologyService );
+                snapshotService, topologyService );
+
+        CoreStateDownloaderService downloadService = new CoreStateDownloaderService( platformModule
```
Make this a LifeCycle and add it to platformModule.life at some appropriate place. Downloads should not be allowed when stopped and stopping could interrupt an on-going download.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/CoreStateDownloaderService.java:
```diff
@@ -0,0 +1,66 @@
+    {
+        if ( currentJob == null || currentJob.hasCompleted() )
+        {
+            synchronized ( this )
```
Just make the entire function synchronized. This double-checking trick is both unsafe and unnecessary.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/PersistentSnapshotDownloader.java:
```diff
@@ -0,0 +1,115 @@
+        state = State.RUNNING;
+        try
+        {
+            applicationProcess.pauseApplier( CoreStateDownloaderService.OPERATION_NAME );
```
This will not work correctly on subsequent attempts after failure because you will pause several times, but only resume once on the last successful attempt. The pause/resume functionality is reference counting so you will be left with a reference count > 0. You need an object with a lifetime of all the attempts to pause and resume, once each.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/PersistentSnapshotDownloader.java:
```diff
@@ -0,0 +1,115 @@
+            {
+                if ( Thread.interrupted() )
+                {
+                    break;
```
Log a warning when interrupted.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/PersistentSnapshotDownloader.java:
```diff
@@ -0,0 +1,115 @@
+                {
+                    log.warn( "No leader found. Retrying in {} ms.", timeout.getMillis() );
+                }
+                LockSupport.parkNanos( TimeUnit.MILLISECONDS.toNanos( timeout.getMillis() ) );
```
Parking cannot be interrupted.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/CoreStateDownloaderService.java:
```diff
@@ -0,0 +1,66 @@
+        this.downloader = downloader;
+        this.applicationProcess = applicationProcess;
+        this.log = logProvider.getLog( getClass() );
+        this.downloadSnapshotGroup = new JobScheduler.Group( "download snapshot", POOLED );
```
I think you should just add this group in the long list in JobScheduler.Groups and reference that.

RagnarW committed:
Respond to feedback from comments

1. Application process always resume.
2. Made CoreStateDownloaderService a LifecycleApplication
3. Added STOPPED state in PersistentSnapshotDownloader. This is called
when LifecycleAdaptor is stopped.
4. Do state check before running. Each PersistentSnapshotDownloader
has run one time only logic.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/CoreStateDownloaderService.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @Override
+    public void stop() throws Throwable
```
synchronize this also

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/CoreStateDownloaderService.java:
```diff
@@ -0,0 +1,66 @@
+    private final CoreStateDownloader downloader;
+    private final CommandApplicationProcess applicationProcess;
+    private final Log log;
+    private PersistentSnapshotDownloader currentJob = null;
```
null is already the default

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/PersistentSnapshotDownloader.java:
```diff
@@ -0,0 +1,157 @@
+        }
+        try
+        {
+            state = State.RUNNING;
```
This variable state is accessed from two threads so you need some type of synchronization mechanism. Might be enough to make it volatile. I think it can be good to split it up into two variables, state and keepRunning, because right now you have two threads updating the same variable and their interactions are hard to follow. The variable state can be updated by PersistentSnaptshotDownloader#run and keepRunning can be updated from #stop. The while-loop then checks keepRunning.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/PersistentSnapshotDownloader.java:
```diff
@@ -0,0 +1,157 @@
+        }
+    }
+
+    private boolean initialStateOk()
```
I don't think you should need this method at all. The state transitions should be obvious and simple. Do you really need INITIATED at all? Can't it just start out in RUNNING-state? Perhaps all you really need is a simple boolean which says isComplete and similarly the method isRunning(), which is only used from tests, can perhaps be eliminated?

RagnarW committed:
Change state behaviour in PersistentDownloader

Moved Stopped state to separate volatile variable keepRunning.

Make state variable volatile.

Remove isRunning method since it is only used in tests.

Synchronize method that move state from initial to running. This
is to make sure only one run is possible so that hasCompleted can be
trusted.

RagnarW committed:
Synchronize stop method in CoreStateDownloaderService

RagnarW committed:
Fix PersistentSnapshotDownloader test
